### DISCUSSION TITLE
Unique Constraint Violation Exception - How to Access Additional Fields That Were Violated

### DISCUSSION THREAD

**Discussion by EvilCheetah** [mikro-orm/mikro-orm#4018] [Q&A] [answered]
# Preface
I am using MikroORM in one of my pet projects in combination with NestJS. It solves all the problems I need, but I have faced some difficulties on Unique Validation. I am trying to rely on the ORM and Database engine for Unique Validation. I have tried to catch an exception and retrieve field(-s), which were violated. However, I am not able to access any property, but the ones in `DriverException`; however, I am able to see additional information when I `console.log | console.table` the error.

I have tried to read the documentation, research the answer in Google, create a [Question in NestJS official Discord](https://discord.com/channels/520622812742811698/1071298586735824966), but found no success doing so.

# Example of Entity I am working with
```ts
import { Entity, PrimaryKey, Property } from "@mikro-orm/core";


@Entity()
export class Country
{
    @PrimaryKey()
    country_id:       number;

    @Property({ unique: true })
    country_name:     string;

    @Property({ unique: true, nullable: true, length: 2 })
    two_alpha_abbr:   string;

    @Property({ unique: true, nullable: true, length: 3 })
    three_alpha_abbr: string;
}
```

# CRUD Logic I have - Create Method
```ts
async create(createCountryDTO: CreateCountryDTO): Promise<Country>
{
    const country = this.countryRepository.create(createCountryDTO);

    try
    {
        await this.countryRepository.persistAndFlush(country);
    }
    catch (error)
    {
        if ( error instanceof UniqueConstraintViolationException )
        {
            console.table(error)
        }
    }

    return country;
}
```

|     (index)      |                   Values                    |
|------------------|---------------------------------------------|
|      length      |                     233                     |
|       name       |    'UniqueConstraintViolationException'     |
|     severity     |                   'ERROR'                   |
|       code       |                   '23505'                   |
|      detail      | 'Key (two_alpha_abbr)=(US) already exists.' |
|       hint       |                  undefined                  |
|     position     |                  undefined                  |
| internalPosition |                  undefined                  |
|  internalQuery   |                  undefined                  |
|      where       |                  undefined                  |
|      schema      |                  'public'                   |
|      table       |                  'country'                  |
|      column      |                  undefined                  |
|     dataType     |                  undefined                  |
|    constraint    |       'country_two_alpha_abbr_unique'       |
|       file       |                'nbtinsert.c'                |
|       line       |                    '664'                    |
|     routine      |             '_bt_check_unique'              |

# Questions
`How can I access the fields provided in the console.table?`
OR
`What is the better approach to the following problem?`

# P.S.
If there is any way I can contribute to the project, I will be happy to do so.

**Comment by B4nan** [comment, marked as answer]
The ORM exception classes are there only so you can detect what kind of error it is, the ORM does not alter the error instance anyhow - it is what your driver produces, just wrapped in a custom class so you can use `instanceof` checks on it. Just cast the error property to anything you want, this should be only type-level issue.

I guess we should try to extract something at least from the most common errors like the unique constraints. I'll try to think about something for v6.

**Reply by EvilCheetah** [reply]
Thanks a lot for it! I really appreciate it! Is there any way I can contribute to that?
For now, I guess, I will rely on manual checks

**Reply by B4nan** [reply]
First step would be to check how the different drivers behave, and define what properties we can get, either directly or with the help of regexps. We need to find a common denominator for the error types. Every driver has an ExceptionConverter class that handles this, e.g. https://github.com/mikro-orm/mikro-orm/blob/master/packages/sqlite/src/SqliteExceptionConverter.ts

So that is where the magic would happen.

**Reply by EvilCheetah** [reply]
I will try to look into it. I will let you know about any updates. I will ping you either here and using any contact provided in your profile.
